### PR TITLE
3DS: Add support for 3DS1 fallback

### PR DESCRIPTION
**Pull request by jlejoux**
Usage changes:

- For 3DS fringerprint fallback to work need to specify the app custom URL scheme during the ProcessOut SDK setup:
```swift
ProcessOut.Setup(projectId: "project-id", urlScheme: "your-app-url-scheme")
```
- Need to pass the UIViewController context to the SDK when requesting a payment (this is needed in order to be able to add the hidden webview to the app layout).
Note the `with: self` parameter
```swift
ProcessOut.makeCardPayment(invoiceId: "invoice-id", token: token, handler: ProcessOut.createThreeDSTestHandler(viewController: self, completion: { (invoiceId, error) in
    if invoiceId != nil {
        print(invoiceId)
        // Authorization successful
    } else {
        print(error)
        // Display error
    }
}), with: self)
```
- New URL callback case to handle when re-opening the app after a web redirect:
```swift
switch processoutResult.type {

{...} // Handle other cases

case .ThreeDSFallbackVerification:
    ProcessOut.continueThreeDSVerification(invoiceId: processoutResult.invoideId, token: processoutResult.value, completion: {(invoiceId, error) in
        // If no error, send the invoice ID to your backend to complete the charge
    })
    break
```

### CODE REVIEW

File: ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                if invoice == "" || token == "" {
+                    self.completion!(nil, nil, ProcessOutException.InternalError)
+                    return
+                }
+                
+                self.completion!(invoice, token, nil)
+            }
+        }
+    }
+    public func webView(_ webView: WKWebView, stop urlSchemeTask: WKURLSchemeTask) {
```
Comment (manuel-huez) [inline ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift:43]: Is that method required?

File: ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                        if queryParams.name == "invoice_id", let value = queryParams.value {
+                            invoice = value
+                        } else if queryParams.name == "token", let value = queryParams.value {
+                            token = value
+                        }
+                    }
+                }
+                
+                if invoice == "" || token == "" {
+                    self.completion!(nil, nil, ProcessOutException.InternalError)
```
Comment (manuel-huez) [inline ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift:35]: You might want to add a friendly message here

File: ProcessOut/Classes/ProcessOut.swift [hunk 3]
```diff
@@ -297,7 +313,7 @@ public class ProcessOut {
     ///   - invoiceId: Invoice generated on your backend
     ///   - token: Card token to be used for the charge
     ///   - handler: Custom 3DS2 handler (please refer to our documentation for this)
-    public static func makeCardPayment(invoiceId: String, token: String, handler: ThreeDSHandler) {
+    public static func makeCardPayment(invoiceId: String, token: String, handler: ThreeDSHandler, with: UIViewController) {
```
Comment (manuel-huez) [inline ProcessOut/Classes/ProcessOut.swift:316]: Does that mean engineers will need to change their code?

File: ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift (hunk 1 above)
Comment (jlejoux) [reply ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift]: Good catch it's not

Comment (jlejoux) [reply ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift:43]: Yes it is required by the WKURLSchemeHandler protocol for this class. We however do not need any action here. I'll add a comment for this

File: ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift (hunk 2 above)
Comment (jlejoux) [reply ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift:35]: Unfortunately, the iOS SDK only support error messages coming from the API. I can however log something in the dev console

File: ProcessOut/Classes/ProcessOut.swift (hunk 3 above)
Comment (jlejoux) [reply ProcessOut/Classes/ProcessOut.swift:316]: If their integration was based on the last release (`makeCardPayment`) yes they will have to add this context

File: ProcessOut/Classes/WebViewReturn.swift [hunk 4]
```diff
@@ -12,16 +12,25 @@ public class WebViewReturn {
     public enum ReturnType {
         case APMAuthorization
         case ThreeDSResult
+        case ThreeDSFallbackVerification
     }
     
     public var success: Bool
     public var value: String = ""
     public var type: ReturnType
+    public var invoideId: String = ""
```
Comment (louis-paul) [inline ProcessOut/Classes/WebViewReturn.swift]:
```suggestion
    public var invoiceId: String = ""
```

File: ProcessOut/Classes/ProcessOut.swift [hunk 5]
```diff
@@ -365,14 +382,96 @@ public class ProcessOut {
...
+                    // Invalid URL
+                    handler.onError(error: ProcessOutException.InternalError)
+                    return
+                }
+                // Prepare the fingerprint hiddenWebview
+                var webView: WKWebView!
+                let preferences = WKPreferences()
+                preferences.javaScriptEnabled = true
+                let configuration = WKWebViewConfiguration()
+                // Check if the device supports custom URL scheme handlind for WebViews
```
Comment (louis-paul) [inline ProcessOut/Classes/ProcessOut.swift]:
```suggestion
                // Check if the device supports custom URL scheme handling for WebViews
```

File: ProcessOut/Classes/ProcessOut.swift [hunk 6]
```diff
@@ -365,14 +382,96 @@ public class ProcessOut {
...
+                    }
+                    configuration.preferences = preferences
+                    // Setup the custom URL scheme handler to detect redirects within the hidden webview
+                    configuration.setURLSchemeHandler(FingerPrintWebViewSchemeHandler(completion: {(invoiceId, token, error) in
+                        // Cancel the timeout as we catched the redirect
+                        timeOutHandler.cancel()
+                        if error != nil {
+                            handler.onError(error: error!)
+                        } else {
+                            // Finggerprint token successfuly received, we continue the authorization flow
```
Comment (louis-paul) [inline ProcessOut/Classes/ProcessOut.swift]:
```suggestion
                            // Fingerprint token successfully received, we continue the authorization flow
```

File: ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift (hunk 2 above)
Comment (manuel-huez) [reply ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift:35]: I'm not sure I understand what you mean by "only support error messages coming from the API". Does it mean we can't return errors with messages when not coming from the ProcessOut API?

File: ProcessOut/Classes/ProcessOut.swift [hunk 7]
```diff
@@ -365,14 +382,96 @@ public class ProcessOut {
...
+        }
+    }
+    
+    private static func fallbackFingerprint(invoiceId: String, URL: String, handler: ThreeDSHandler, with: UIViewController) {
+        do {
+            let miscGatewayRequest = MiscGatewayRequest(fingerprintResponse: "{\"threeDS2FingerprintTimeout\":true}")
+            miscGatewayRequest.headers = ["Content-Type": "application/json"]
+            miscGatewayRequest.url = URL
+            let encodedJson = try JSONEncoder().encode(miscGatewayRequest)
+            if let base64Encoded = String(data: encodedJson.base64EncodedData(), encoding: .utf8) {
```
Comment (manuel-huez) [inline ProcessOut/Classes/ProcessOut.swift]: Perhaps adding a helper in this `MiscGatewayRequest` object would be a good idea as well, similarly to what we've done on Android

File: ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift (hunk 2 above)
Comment (jlejoux) [reply ProcessOut/Classes/FingerPrintWebViewSchemeHandler.swift:35]: Yep. For now we only have a few different error "types" being:
```swift
public enum ProcessOutException: Error {
    case NetworkError
    case MissingProjectId
    case BadRequest(errorMessage: String, errorCode: String)
    case InternalError
    case GenericError(error: Error)
}
```

I don't want to add custom messages to the `BadRequest` case as I want to keep `errorCode` coming from the API/documentation.

We could add a custom case but merchants will need to update their integration to support it.

**Review by louis-paul — CHANGES_REQUESTED** [review]
lgtm modulo typos

**Review by louis-paul — APPROVED** [review]

**Pull request merged**